maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/StatelessXmlReporter.java

```diff
@@ -456,21 +456,13 @@ private static void getTestProblems(
             }
         }
 
-        boolean hasNestedElements = createOutErrElementsInside & stackTrace != null;
-
-        if (stackTrace != null) {
-            if (hasNestedElements) {
-                ppw.startElement("stackTrace");
+        if (enableNestedOutErrElements) {
+            ppw.startElement("stackTrace");
+            if (stackTrace != null) {
+                extraEscapeElementValue(stackTrace, outputStreamWriter, ppw, fw);
             }
+            ppw.endElement();
 
-            extraEscapeElementValue(stackTrace, outputStreamWriter, ppw, fw);
```

Comment:
This still needs to be called if `stackTrace != null` and `!enableNestedOutErrElements`. Otherwise, `error` and `failure` elements won't contain the stack trace as their content.

Reply:
I think I understand what you mean. Let me rephrase and I guess you are right here, the schema is poorly designed. According to https://maven.apache.org/surefire/maven-surefire-plugin/xsd/surefire-test-report.xsd both `failure` and `error` aren't complex elements, thus do *not* contain `stackTrace`. Their element body contains the strack trace?

Reply:
Yes, that's right.

## Before (with 3.2.3 and earlier)
```xml
<testcase name="addsTwoNumbers" classname="com.example.project.CalculatorTests" time="0.014">
  <failure message="1 + 1 should equal 2 ==&gt; expected: &lt;3&gt; but was: &lt;2&gt;" type="org.opentest4j.AssertionFailedError"><![CDATA[org.opentest4j.AssertionFailedError: 1 + 1 should equal 2 ==> expected: <3> but was: <2>
  at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
  at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
  at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
  at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
  at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:563)
  at com.example.project.CalculatorTests.addsTwoNumbers(CalculatorTests.java:26)
  at java.base/java.lang.reflect.Method.invoke(Method.java:578)
  at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
  at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
]]></failure>
</testcase>
```

## After this change
```xml
<testcase name="addsTwoNumbers" classname="com.example.project.CalculatorTests" time="0.014">
  <failure message="1 + 1 should equal 2 ==&gt; expected: &lt;3&gt; but was: &lt;2&gt;" type="org.opentest4j.AssertionFailedError"/>
</testcase>
```

Reply:
Thanks for the confirmation. I will cancel the vote today, create an improvement request to change the schema in the next major to a complex element with `stackTrace` to align with the rest and another one to fix this regression. I will ping you as soon as I have something to test.

Reply:
Vote canceled.

Reply:
Fix: https://github.com/apache/maven-surefire/pull/709